crates/dekoder/src/inventory.rs

```diff
@@ -0,0 +1,87 @@
+//! Assets manager.
+use glob::glob;
+use hashbrown::HashMap;
+use std::ffi::OsString;
+use std::fs::read_to_string;
+use std::path::PathBuf;
+use yaml_rust2::{Yaml, YamlLoader};
+
+use crate::{EKOError, Result};
+
+/// Headers are in yaml files.
+const HEADER_EXT: &'static str = "*.yaml";
+
+/// Header type in an inventory.
+pub(crate) trait HeaderT {
```

Comment:
I'd remove this `trait` completely and its use with `From<&Yaml>` and `Eq` which are the standard traits. For ULPS I'd use a `const` that's reasonably large, for instance `64`.

Reply:
I made an attempt in [f0d6a27](https://github.com/NNPDF/eko/pull/400/commits/f0d6a276360d1b61736e4e259f77036148b9dbd8), but it is not working since I don't fully understand Rust yet :innocent: in particular there is a problem with lifetimes: in particular here
https://github.com/NNPDF/eko/blob/f0d6a276360d1b61736e4e259f77036148b9dbd8/crates/dekoder/src/inventory.rs#L50
Rust doesn't know that the error has the correct type and in particular he can not know. The exact error is
```
error[E0277]: `?` couldn't convert the error to `EKOError`
  --> crates/dekoder/src/inventory.rs:50:37
   |
50 |                 K::try_from(&cnt[0])?,
   |                                     ^ the trait `std::convert::From<<K as std::convert::TryFrom<&yaml_rust2::Yaml>>::Error>` is not implemented for `EKOError`, which is required by `std::result::Result<(), EKOError>: std::ops::FromResidual<std::result::Result<std::convert::Infallible, <K as std::convert::TryFrom<&yaml_rust2::Yaml>>::Error>>`
   |
   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
   = note: required for `std::result::Result<(), EKOError>` to implement `std::ops::FromResidual<std::result::Result<std::convert::Infallible, <K as std::convert::TryFrom<&yaml_rust2::Yaml>>::Error>>`
help: consider introducing a `where` clause, but there might be an alternative better way to express this requirement
   |
30 | impl<K: Eq + for<'a> TryFrom<&'a Yaml>> Inventory<K> where EKOError: std::convert::From<<K as std::convert::TryFrom<&yaml_rust2::Yaml>>::Error> {
   |                                                      ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

For more information about this error, try `rustc --explain E0277`.
error: could not compile `dekoder` (lib test) due to 1 previous error
```
however the suggested fix doesn't work, because there is again a reference appearing in the where clause and I don't know how to fix that. (I understand the general idea of lifetimes, i.e. `Inventory::load_keys` is promising to `EvolutionPoint::try_from` that it will hold the yaml data the whole time.) Note that we need `TryFrom` instead of `From` since reading yaml is a complicated action.

Can you give me some hint?

Reply:
Try changing `TryFrom<&'a Yaml>` to `TryFrom<&'a Yaml, Error = EKOError>`.

Reply:
yes! done in 599fb855 - such an easy fix :upside_down_face: a) why didn't the compiler already suggested that and b) I did look to [the documentation](https://doc.rust-lang.org/std/convert/trait.TryFrom.html) but there I read it as it only takes one generic argument; can I just specify "the member" `Error` as well?

Reply:
`Error` is called an ['associated type'](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#specifying-placeholder-types-in-trait-definitions-with-associated-types) of `TryFrom` and may be specified if you need to restrict it, otherwise you can leave it out and it is inferred by the compiler. In this case you had to specify it because otherwise the error type of the return type of `try_from` is unknown and it must be at least convertible into an `EKOError` when it is called in `Inventory::load_keys`; that's what the compiler message said. The suggested `where` clause of the compiler is much less strict (`Error` convertible into `EKOError`) than my suggestion (`Error` IS `EKOError`), so that's fine.